### PR TITLE
Fix grammar verification

### PR DESCRIPTION
**Pull request by dns2utf8**
* Use make check-lexer to verify the grammar.
 * Extend grammar/README
 * Add make clean-grammar rule
 * Add target check-build-lexer-verifier to make tidy, so it will build the verifier with every build and catch future errors

This is the continuation of #34994

r? @steveklabnik @jonathandturner @alexcrichton

### CODE REVIEW

File: mk/grammar.mk [hunk 1]
```diff
@@ -37,7 +37,7 @@ $(BG):
 
 $(BG)RustLexer.class: $(BG) $(SG)RustLexer.g4
 	$(Q)$(CFG_ANTLR4) -o $(BG) $(SG)RustLexer.g4
-	$(Q)$(CFG_JAVAC) -d $(BG) $(BG)RustLexer.java
+	$(Q)$(CFG_JAVAC) -d $(BG) -classpath /usr/share/java/antlr-complete.jar $(BG)RustLexer.java
```
Comment (alexcrichton) [inline mk/grammar.mk]: Would it be possible to abstract this to avoid hardcoding it?

**Review by alexcrichton — COMMENTED** [review]
Nice! Hopefully we'll be able to easily get a "bot" up and running for this soon once we switch to Travis

File: mk/grammar.mk (hunk 1 above)
Comment (dns2utf8) [inline mk/grammar.mk]: Do you mean something like this?
```
echo $(find /usr/ -name antlr-complete.jar 2>/dev/null)
```

Comment (alexcrichton) [inline mk/grammar.mk]: I was thinking more of a ./configure style thing (or detection elsewhere)

**Comment by steveklabnik** [comment]
Cool!

**Comment by alexcrichton** [comment]
@dns2utf8 thoughts about configuring where antlr itself is located given [my previous comment](https://github.com/rust-lang/rust/pull/37607#discussion_r86690550)?

File: src/grammar/check.sh [hunk 2]
```diff
@@ -20,11 +20,11 @@ skipped=0
 check() {
     grep --silent "// ignore-lexer-test" "$1";
 
-    # if it's *not* found...
+    # if it is *not* found...
```
Comment (GuillaumeGomez) [inline src/grammar/check.sh]: Not sure to see the usefulness of this change...

File: src/grammar/check.sh [hunk 3]
```diff
@@ -20,11 +20,11 @@ skipped=0
...
     if [ $? -eq 1 ]; then
-        cd $2 # This `cd` is so java will pick up RustLexer.class. I couldn't
+        cd $2 # This `cd` is so java will pick up RustLexer.class. I could not
```
Comment (GuillaumeGomez) [inline src/grammar/check.sh]: Same.

File: src/grammar/check.sh [hunk 4]
```diff
@@ -20,11 +20,11 @@ skipped=0
...
         # figure out how to wrangle the CLASSPATH, just adding build/grammar
-        # didn't seem to have any effect.
+        # did not seem to have any effect.
```
Comment (GuillaumeGomez) [inline src/grammar/check.sh]: Same.

**Comment by dns2utf8** [comment]
@alexcrichton I have never worked with the configure-step. I am looking into it now.

File: src/grammar/check.sh (hunk 2 above)
Comment (dns2utf8) [inline src/grammar/check.sh]: Well I read it and I thought I change it. Is it wrong?

File: configure [hunk 5]
```diff
@@ -852,6 +852,7 @@ probe_need CFG_CMAKE cmake
 # probe for it only in this case.
 if [ -n "$CFG_ANTLR4" ]
 then
+   putvar "CFG_ANTLR4_JAR                     $(echo -n $(find /usr/ -name antlr-complete.jar 2>/dev/null | head -n 1))"
```
Comment (dns2utf8) [inline configure]: I have a strange bug: With this line i save the `CFG_ANTLR4_JAR` to the generated config but it also generates an additional line after this entry. 
That entry is empty and stops the use of the `config.mk`.

Do you have an idea?

File: src/grammar/check.sh (hunk 2 above)
Comment (GuillaumeGomez) [inline src/grammar/check.sh]: Well it's not wrong, but "it's is just as good as "it is" so why changing it?

Comment (dns2utf8) [inline src/grammar/check.sh]: I learned it that way in school, an old habit.

File: configure (hunk 5 above)
Comment (alexcrichton) [inline configure]: Is that perhaps because the quote after `JAR` isn't terminated?

**Comment by dns2utf8** [comment]
I extended the find to search the users homedirectory for `antlr-complete.jar` if it is installed on the system. 
I do not have access to a mac, but I hope this will work there too.

**Pull request merged**